### PR TITLE
Sync payload refactor

### PR DESCRIPTION
**Pull request by proteusvacuum**
Breaking the `get_payload` method of `CleanOwnerCaseSyncOperation` into its own class. 

Hopefully this should make it easier to profile individual parts of the sync operation, since there has been some issues with workers with a large number of cases (also every time I looked at the previous method I got a bit confused since it was doing so much). 

Not sure if this was the right way of doing this, but figured I'd throw this up to get some comments on it.
@snopoke @czue 
cc: @emord

### CODE REVIEW

File: corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py [hunk 1]
```diff
@@ -23,6 +22,145 @@ def get_case_payload(restore_state):
...
+        for update in updates:
+            case = update.case
+            self.potential_elements_to_sync[case.case_id] = get_elements_for_response(update, self.restore_state)
+            self._process_case_update(case)
+            self.checked_cases.add(case.case_id)  # mark case as checked
+
+    def _add_commtrack_elements_to_response(self, updates):
+        # commtrack ledger sections for this batch
+        commtrack_elements = get_stock_payload(
+                self.restore_state.project, self.restore_state.stock_settings,
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py]: E126 continuation line over-indented for hanging indent

File: corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py [hunk 2]
```diff
@@ -23,6 +22,145 @@ def get_case_payload(restore_state):
...
+                [CaseStub(update.case.case_id, update.case.type) for update in updates]
+            )
```
Comment (dimagimon) [inline corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py]: E121 continuation line under-indented for hanging indent

File: corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py [hunk 3]
```diff
@@ -23,6 +22,145 @@ def get_case_payload(restore_state):
...
+        self.update_case_ids_on_phone()
+        self.move_no_longer_owned_cases_to_dependent_list_if_necessary()
+        irrelevant_cases = self.purge_and_get_irrelevant_cases()
+        self.compile_response(irrelevant_cases)
+
+        return self.response
+
+    def _get_next_case_batch(self):
+        ids = pop_ids(self.case_ids_to_sync, chunk_size)
+        # TODO: see if we can avoid wrapping - serialization depends on it heavily for now
```
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py]: Don't think this is really relevant anymore now that we have SQL and Couch cases. If we didn't wrap couch cases we'd have to convert SQL cases to JSON which is probably as much effort.

File: corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py [hunk 4]
```diff
@@ -23,6 +22,145 @@ def get_case_payload(restore_state):
...
+        return [
+            case for case in self.case_accessor.get_cases(ids)
+            if not case.is_deleted and case_needs_to_sync(case, last_sync_log=self.restore_state.last_sync_log)
+        ]
+
+    def process_case_batch(self, case_batch):
+        updates = get_case_sync_updates(
+            self.restore_state.domain, case_batch, self.restore_state.last_sync_log
+        )
+        self._add_commtrack_elements_to_response(updates)
```
Comment (snopoke) [inline corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py]: I realize you didn't change this but just wondering if it's correct to add commtrack elements for all these cases if we may remove some of them later in `compile_response`

**Comment by snopoke** [comment]
Very nice refactor. Build fail real:
```
======================================================================
FAIL: test_Broken_Extension_Chain (casexml.apps.phone.tests.test_extension_indexes.IndexTreeTest)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/mnt/corehq/util/test_utils.py", line 232, in inner
    return helper(*args, **kwargs)
  File "/mnt/corehq/util/test_utils.py", line 205, in __call__
    call_with_settings(fn_with_pre_and_post, run_config.settings, args, kwargs)
  File "/mnt/corehq/util/test_utils.py", line 220, in call_with_settings
    fn(*args, **kwargs)
  File "/mnt/corehq/util/test_utils.py", line 201, in fn_with_pre_and_post
    self.fn(*args, **kwargs)
  File "/mnt/corehq/util/test_utils.py", line 87, in wrapper
    return func(*args, **kw)
  File "corehq/ex-submodules/casexml/apps/phone/tests/test_extension_indexes.py", line 39, in test
    assert_user_doesnt_have_cases(self, self.user, undesired_cases)
  File "corehq/ex-submodules/casexml/apps/case/tests/util.py", line 123, in assert_user_doesnt_have_cases
    should_have=False, **kwargs)
  File "corehq/ex-submodules/casexml/apps/case/tests/util.py", line 193, in check_user_has_case
    matches = [check_block(case_block) for case_block in case_blocks]
  File "corehq/ex-submodules/casexml/apps/case/tests/util.py", line 185, in check_block
    "but shouldn't:%s" % (user.username, case_id, extra_info())
AssertionError: User 'syncguy' gets case 'b' but shouldn't:
<ns0:case xmlns:ns0="http://commcarehq.org/case/transaction/v2" case_id="b" date_modified="2016-06-10T01:08:32.959712Z" />
['<ns0:case xmlns:ns0="http://commcarehq.org/case/transaction/v2" case_id="a" date_modified="2016-06-10T01:08:31.794765Z" user_id="68e6e674fe69024180ece0380057023b"><ns0:create><ns0:case_type>mother</ns0:case_type><ns0:case_name /><ns0:owner_id>68e6e674fe69024180ece0380057023b</ns0:owner_id></ns0:create></ns0:case>', '<ns0:case xmlns:ns0="http://commcarehq.org/case/transaction/v2" case_id="c" date_modified="2016-06-10T01:08:31.794903Z" user_id="68e6e674fe69024180ece0380057023b"><ns0:create><ns0:case_type>mother</ns0:case_type><ns0:case_name /><ns0:owner_id>943737b1fff64ccd9cd91a53e7cda01a</ns0:owner_id></ns0:create><ns0:index><ns0:extension_1 case_type="default_related_case_type" relationship="extension">b</ns0:extension_1></ns0:index></ns0:case>', '<ns0:case xmlns:ns0="http://commcarehq.org/case/transaction/v2" case_id="b" date_modified="2016-06-10T01:08:31.795112Z" user_id="68e6e674fe69024180ece0380057023b"><ns0:create><ns0:case_type>mother</ns0:case_type><ns0:case_name /><ns0:owner_id>95cccf9c2b1942c8b5516c333debcf60</ns0:owner_id></ns0:create><ns0:index><ns0:extension_0 case_type="default_related_case_type" relationship="extension">a</ns0:extension_0></ns0:index><ns0:close /></ns0:case>']
-------------------- >> begin captured stdout << ---------------------
<function test at 0x7f25f6f13e60> failed with the following settings:
settings.TESTS_SHOULD_USE_SQL_BACKEND = False
```

File: corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py (hunk 4 above)
Comment (proteusvacuum) [inline corehq/ex-submodules/casexml/apps/phone/data_providers/case/clean_owners.py]: made an attempt here: https://github.com/dimagi/commcare-hq/pull/11995/commits/4864e59cad98e0f74834c4e8f427851f110458e1

**Comment by proteusvacuum** [comment]
:white_check_mark:

**Comment by snopoke** [comment]
Nice, looks great. Not merging due to label.

**Comment by proteusvacuum** [comment]
Removed label. Not sure if @czue wanted to take a look first, otherwise I'm happy with this.

**Comment by czue** [comment]
I'll take a look eventually but no need to block merge on my review

**Comment by czue** [comment]
nice this looks like a great set of changes. thanks!

**Pull request merged**